### DISCUSSION TITLE
BaseOperator: method when user sets task as failed?

### DISCUSSION THREAD

**Discussion by ecodina** [apache/airflow#37975] [Q&A] [answered]
Hello,
I'm developing a custom operator. I would like to execute some code whenever a user sets the state as failed. For example: if a task is RUNNING (or DEFERRED), and then the user sets it failed, I'd like to run some custom code, which would tidy up some things.

I've been reading the BaseOperator file, and I've found two methods which I could probably override and may be useful:

- `__setstate__`
- `on_kill`

Is anyone familiar with this issue? Thank you very much!

**Comment by potiuk** [comment, marked as answer]
I don't think BaseOperator is foreseeen for such a way of extending it. Anything you implement now by overriding private methods is likely not going to survive future updates and you should not implement it this way. Also on_kill is not going to work in deferrable operator case.

The only places you can `interact` with Base Operator are basically __init__ and `execute` methods - those are foreseen for extensions. There are multiple ways state change might happen - for example when you clean a task from UI - your BaseOperator class running implementation will not even be aware of that. The task's state database will be changed by the webserver and the next thing that happens is the monitoring process that runs in the worker will see the state change and your task will get stopped (for example). Which might or might not cause on_kill method. If your state is in deferred state, then on_kill method is not called for example and the BaseOperator class of yours is not even instantiated, so it's impossible to handle it in the operator class.

I think your best bet is to use listeners and TaskInstanceState  changes events if you want to do some cleanups when state of task instance changes https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/listeners.html#taskinstance-state-change-events.

You will have to implement it as airflow Plugin. And it's fairly recent addition to Airflow so you might need to upgrade to latest version (which is a good thing in general regardless).

**Reply by ecodina** [reply]
Thanks, @potiuk for your answer and for pointing me to a better solution!
